File: alien/java/src/CountXMLNodes.java
import javax.xml.stream.*;
import javax.xml.stream.events.StartElement;
import java.io.FileInputStream;
import java.io.BufferedInputStream;
import java.util.*;

public class CountXMLNodes {

    public static Map<String, Integer> countNodes(String xmlFilePath, Set<String> filterNodes) throws Exception {
        Map<String, Integer> nodeCounts = new HashMap<>();
        XMLInputFactory factory = XMLInputFactory.newInstance();
        FileInputStream fis = new FileInputStream(xmlFilePath);
        BufferedInputStream bis = new BufferedInputStream(fis, 8192 * 1024);
        XMLStreamReader reader = factory.createXMLStreamReader(bis);

        while (reader.hasNext()) {
            int event = reader.next();
            if (event == XMLStreamConstants.START_ELEMENT) {
                String nodeName = reader.getLocalName();
                if (filterNodes == null || filterNodes.contains(nodeName)) {
                    nodeCounts.put(nodeName, nodeCounts.getOrDefault(nodeName, 0) + 1);
                }
            }
        }
        reader.close();
        fis.close();
        return nodeCounts;
    }

    public static void main(String[] args) {
        if (args.length < 1) {
            System.err.println("Usage: java XmlNodeCounter <path_to_xml_file> [node_name1 node_name2 ...]");
            System.exit(1);
        }

        String xmlFilePath = args[0];
        Set<String> filterNodes = null;
        if (args.length > 1) {
            filterNodes = new HashSet<>(Arrays.asList(Arrays.copyOfRange(args, 1, args.length)));
        }

        try {
            Map<String, Integer> nodeCounts = countNodes(xmlFilePath, filterNodes);
            System.out.println("Node counts:");
            for (Map.Entry<String, Integer> entry : nodeCounts.entrySet()) {
                System.out.println(entry.getKey() + ": " + entry.getValue());
            }
        } catch (Exception e) {
            System.err.println("Error processing XML file: " + e.getMessage());
            System.exit(1);
        }
    }
}

File: alien/Scala/src/CountXmlNodes.scala
import javax.xml.stream.{XMLInputFactory, XMLStreamConstants}
import java.io.{FileInputStream, BufferedInputStream}
import scala.collection.mutable

object CountXMLNodes {
  def countNodes(
      xmlFilePath: String,
      filterNodes: Set[String]
  ): Map[String, Int] = {
    val nodeCounts = mutable.Map[String, Int]()
    val factory = XMLInputFactory.newInstance()
    val fis = new FileInputStream(xmlFilePath)
    val bis = new BufferedInputStream(fis, 64 * 1024)
    val reader = factory.createXMLStreamReader(bis)

    try {
      while (reader.hasNext) {
        val event = reader.next()
        if (event == XMLStreamConstants.START_ELEMENT) {
          val nodeName = reader.getLocalName
          if (filterNodes.isEmpty || filterNodes.contains(nodeName)) {
            nodeCounts(nodeName) = nodeCounts.getOrElse(nodeName, 0) + 1
          }
        }
      }
    } finally {
      reader.close()
      bis.close()
      fis.close()
    }
    nodeCounts.toMap
  }

  def main(args: Array[String]): Unit = {
    if (args.length < 1) {
      Console.err.println(
        "Usage: scala CountXMLNodes <path_to_xml_file> [node_name1 node_name2 ...]"
      )
      sys.exit(1)
    }

    val xmlFilePath = args(0)
    val filterNodes =
      if (args.length > 1) args.drop(1).toSet
      else Set.empty[String]

    try {
      val nodeCounts = countNodes(xmlFilePath, filterNodes)
      println("Node counts:")
      nodeCounts.foreach { case (node, count) =>
        println(s"$node: $count")
      }
    } catch {
      case e: Exception =>
        Console.err.println(s"Error processing XML file: ${e.getMessage}")
        sys.exit(1)
    }
  }
}
